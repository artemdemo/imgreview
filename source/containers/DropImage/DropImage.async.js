import lazify from '../lazify';

export default lazify(() => import(/* webpackChunkName: "<DropImage />" */ './DropImage'));
