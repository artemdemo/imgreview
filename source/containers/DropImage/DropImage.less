.drop-image {
    min-height: 300px;
    padding: 10px;
    border: 2px dashed lightgray;
    width: 80%;
    margin: 10px auto 0;
    position: relative;

    &::before {
        content: 'Drop image here';
        top: 50%;
        left: 50%;
        position: absolute;
        transform: translate(-50%, -50%);
        color: gray;
    }
}
    .drop-image_has-image {
        margin: 0;
        width: auto;
        border: none;
        padding: initial;
        min-height: auto;
        position: initial;

        &::before {
            display: none;
        }
    }
    .drop-image_active {
        border-color: lightblue;
    }
