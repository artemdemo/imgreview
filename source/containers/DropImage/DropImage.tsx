import React from 'react';
import classnames from 'classnames';
import Dropzone from 'react-dropzone';
import { connect } from 'react-redux';
import { createGlobalStyle } from 'styled-components';
import loadImage from '../../services/loadImage';
import * as canvasApi from '../../../srcCanvas/api';
import { TReduxState } from '../../reducers';
import { TStateCanvas } from '../../model/canvas/canvasReducer';
import { TAddImage, addImage } from '../../model/canvas/canvasActions';

type Props = {
    canvas: TStateCanvas,
    addImage: TAddImage,
};

const DropzoneCss = createGlobalStyle`
    .drop-image {
        min-height: 300px;
        padding: 10px;
        border: 2px dashed lightgray;
        width: 80%;
        margin: 10px auto 0;
        position: relative;
        outline: none;
    
        &::before {
            content: 'Drop image here';
            top: 50%;
            left: 50%;
            position: absolute;
            transform: translate(-50%, -50%);
            color: gray;
        }
    }
        .drop-image_has-image {
            margin: 0;
            width: auto;
            border: none;
            padding: initial;
            min-height: auto;
            position: initial;
    
            &::before {
                display: none;
            }
        }
        .drop-image_active {
            border-color: lightblue;
        }
`;

// @docs https://react-dropzone.netlify.com/#proptypes
//
class DropImage extends React.PureComponent<Props> {
    onImageLoaded = (file, data) => {
        const { addImage } = this.props;
        canvasApi.setImage(data);
        addImage({
            name: file.name,
        });
    };

    onDrop = (files) => {
        const file = files[0];
        if (file) {
            loadImage(file)
                .then(this.onImageLoaded.bind(null, file));
        }
    };

    renderDropZone = (propsZone) => {
        const { getRootProps } = propsZone;
        const { canvas } = this.props;
        const dropImageClass = classnames({
            'drop-image': true,
            'drop-image_has-image': canvas.imageHeight !== 0 || canvas.imageWidth !== 0,
            'drop-image_active': propsZone.isDragActive,
        });
        return (
            <div
                {...getRootProps()}
                className={dropImageClass}
            >
                {this.props.children}
            </div>
        );
    };

    render() {
        return (
            <React.Fragment>
                <DropzoneCss />
                <Dropzone
                    onDrop={this.onDrop}
                >
                    {this.renderDropZone}
                </Dropzone>
            </React.Fragment>
        );
    }
}

export default connect(
    (state: TReduxState) => ({
        canvas: state.canvas,
    }), {
        addImage,
    },
)(DropImage);
