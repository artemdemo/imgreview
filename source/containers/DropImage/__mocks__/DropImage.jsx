import React from 'react';

const DropImage = props => (
    <div data-mock='DropImage'>
        {props.children}
    </div>
);

export default DropImage;
