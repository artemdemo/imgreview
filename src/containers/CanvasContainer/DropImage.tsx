import React, { useContext } from 'react';
import Dropzone from 'react-dropzone';
import loadImage from '../../services/loadImage';
import * as gaService from '../../services/ganalytics';
import { AppStateContext } from '../../model/AppStateContext';
import { showNotification } from '../../model/notifications/notificationsActions';
import { t } from '../../services/i18n';
import { NotificationType } from '../../model/notifications/Notification';

// @docs https://react-dropzone.netlify.com/#proptypes
//
export const DropImage: React.FC = (props) => {
  const { children } = props;
  const {
    state: {
      menu,
      canvas: { canvasApi },
    },
    dispatch,
  } = useContext(AppStateContext);

  const onDrop = async (files: File[], _: any[], e: any) => {
    if (files.length > 0) {
      for (const file of files) {
        if (!file.type.startsWith('image')) {
          dispatch(
            showNotification({
              message: t('image-type-error'),
              type: NotificationType.Error,
            }),
          );
          continue;
        }
        const data = await loadImage(file);
        canvasApi?.blurShapes();
        canvasApi?.setImage({
          image: data.image,
          name: data.name,
          pos: {
            x: e.clientX - data.image.width / 2,
            y: e.clientY - menu.menuHeight - data.image.height / 2,
          },
        });
      }
    }
    gaService.sendEvent({
      eventCategory: gaService.EEventCategories.GlobalInteraction,
      eventAction: gaService.EEventActions.DropImage,
    });
  };

  return (
    <Dropzone onDrop={onDrop}>
      {(propsZone) => {
        const { getRootProps } = propsZone;
        return <div {...getRootProps()}>{children}</div>;
      }}
    </Dropzone>
  );
};
