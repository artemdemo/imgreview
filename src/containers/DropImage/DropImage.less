.drop-image {
  min-height: 300px;
  padding: 10px;
  border: 2px dashed lightgray;
  position: absolute;
  outline: none;
  left: 10px;
  right: 10px;
  z-index: 1;

  &::before {
    content: 'Drop image here';
    top: 50%;
    left: 50%;
    position: absolute;
    transform: translate(-50%, -50%);
    color: gray;
    font-family: Lato, Arial, serif;
  }
}
.drop-image_has-image {
  margin: 0;
  width: auto;
  border: none;
  padding: initial;
  min-height: auto;
  position: initial;

  &::before {
    display: none;
  }
}
.drop-image_active {
  border-color: lightblue;
}
