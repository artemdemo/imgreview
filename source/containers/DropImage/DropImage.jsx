import React from 'react';
import classnames from 'classnames';
import _isFunction from 'lodash/isFunction';
import DropzonePkg from 'react-dropzone';
import { connect } from 'react-redux';
import loadImage from '../../services/loadImage';
import { addImage } from '../../model/canvas/canvasActions';
import CanvasImage from '../../canvas/CanvasImage';

import './DropImage.less';

// There is breaking change in webpack 4
// @link https://medium.com/webpack/webpack-4-import-and-commonjs-d619d626b655
const Dropzone = _isFunction(DropzonePkg) ? DropzonePkg : DropzonePkg.default;

// @docs https://react-dropzone.netlify.com/#proptypes
//
class DropImage extends React.PureComponent {
    onDrop = (files) => {
        const file = files[0];
        const { canvas, addImage } = this.props;
        if (file) {
            loadImage(file)
                .then(({ image, name }) => {
                    // ToDo: this code is duplicate from OpenImageDialog.jsx
                    if (canvas.image) {
                        canvas.image.destroy();
                    }
                    canvas.stage.setAttr('width', image.width);
                    canvas.stage.setAttr('height', image.height);
                    const canvasImage = new CanvasImage({
                        image,
                    });
                    canvasImage.addToStage(canvas.stage);
                    addImage(canvasImage, name);
                });
        }
    };

    render() {
        const { canvas } = this.props;
        const dropImageClass = classnames({
            'drop-image': true,
            'drop-image_has-image': canvas.image !== null,
        });
        return (
            <Dropzone
                className={dropImageClass}
                activeClassName='drop-image_active'
                onDrop={this.onDrop}
                disableClick
            >
                {this.props.children}
            </Dropzone>
        );
    }
}

export default connect(
    state => ({
        canvas: state.canvas,
    }), {
        addImage,
    },
)(DropImage);
