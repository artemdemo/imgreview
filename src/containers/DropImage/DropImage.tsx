import React from 'react';
import classnames from 'classnames';
import Dropzone from 'react-dropzone';
import {useDispatch, useSelector} from 'react-redux';
import loadImage, { LoadImageResult } from '../../services/loadImage';
import { TReduxState } from '../../reducers';
import { TStateCanvas } from '../../model/canvas/canvasReducer';
import { addImage } from '../../model/canvas/canvasActions';
import * as gaService from '../../services/ganalytics';
import './DropImage.less';

// @docs https://react-dropzone.netlify.com/#proptypes
//
export const DropImage: React.FC = (props) => {
  const { children } = props;
  const dispatch = useDispatch();
  const canvas = useSelector<TReduxState, TStateCanvas>((state) => state.canvas);

  const onDrop = async (files: File[]) => {
    const file = files[0];
    if (file) {
      const data = await loadImage(file);
      dispatch(addImage({
        image: data.image,
        name: data.name,
      }));
    }
    gaService.sendEvent({
      eventCategory: gaService.EEventCategories.GlobalInteraction,
      eventAction: gaService.EEventActions.DropImage,
    });
  };

  return (
    <Dropzone onDrop={onDrop}>
      {(propsZone) => {
        const { getRootProps } = propsZone;
        const dropImageClass = classnames({
          'drop-image': true,
          'drop-image_has-image': canvas.height !== 0 || canvas.width !== 0,
          'drop-image_active': propsZone.isDragActive,
        });
        return (
          <div {...getRootProps()} className={dropImageClass}>
            {children}
          </div>
        );
      }}
    </Dropzone>
  );
};
