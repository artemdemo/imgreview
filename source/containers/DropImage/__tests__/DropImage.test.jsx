import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import DropImage from '../DropImage';

jest.mock('../../../services/loadImage');

const state = {
    canvas: {
        image: {},
        stage: {
            setAttr() {},
        },
    },
};

describe('DropImage', () => {
    jest.clearAllMocks();
    const reactReduxMock = require('react-redux');

    it('should render with image', () => {
        const tree = renderer.create(
            <DropImage canvas={state.canvas} />,
        ).toJSON();

        expect(reactReduxMock.__getLastMaps().mapStateToProps(state)).toEqual({
            canvas: state.canvas,
        });
        expect(tree).toMatchSnapshot();
    });

    it('should render without image', () => {
        const tree = renderer.create(
            <DropImage
                canvas={{
                    ...state.canvas,
                    image: null,
                }}
            />,
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should call onDrop', () => {
        const loadImage = require('../../../services/loadImage');
        const wrapper = mount(
            <DropImage
                canvas={{
                    ...state.canvas,
                    image: null,
                }}
            />,
        );
        wrapper.find('div').simulate('click');
        expect(loadImage.default).toBeCalledWith(
            {data: 'mock file'}
        );
    });
});
