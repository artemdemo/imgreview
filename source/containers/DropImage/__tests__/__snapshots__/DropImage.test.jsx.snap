// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropImage should render with image 1`] = `
<div
  className="drop-image drop-image_has-image"
/>
`;

exports[`DropImage should render without image 1`] = `
<div
  className="drop-image"
/>
`;
